### DISCUSSION TITLE
Proposal: Support for integration controlled "extra" attributes for entries in the device and entity registry

### DISCUSSION THREAD

**Discussion by raman325** [home-assistant/architecture#746] [General]
## Context

In the core we have generally tried to avoid recording static attributes about an entity in the state machine, which makes sense, because if they are static, it's wasted overhead to maintain them there. But there are places where extra attributes which aren't common across all integrations would be useful. Examples from `zwave_js`:

Device attributes:
- Node ID

Entity attributes:
- primary value ID
- command class
- property
- property key
- for notification binary sensors, state key/value

## Proposal

1. Extend the `RegistryEntry` model for both device and entity registries to have a new dictionary property (e.g. `extra_attributes`) where integrations can provide key value pairs.
2. Update the Entity model to support providing this information in the entity class like we do for `device_info`.
3. Update `device_attr` and `is_device_attr` templates to support pulling attributes from this new dictionary
4. Add `entity_attr` and `is_entity_attr` templates to support pulling attributes from the entity registry entry (names may need to change to avoid confusion with state attributes)
5. Update the frontend to support displaying these properties on the device panel/entity modal (for `zwave_js` we are already doing this in a custom panel)

## Consequences

1. If entities need to update these static properties after initial registry entry creation, they will need to explicitly update the registry like they have to do today for e.g. the device registry. There may be a way to design around this but that wouldn't work with our current architecture
2. Frontend changes would have to be done by someone else and we'd have to think through how translations would work for these properties (if needed)
3. Entity registry entry attributes vs state attributes might be confusing to users. One funky option would be to add the registry entry attributes to the state attributes so they can be accessed from the same place (e.g. with templates, from the UI, etc.) but we would want to do it in a way to avoid writing it to the state machine.

**Comment by balloob** [comment]
For entities we already store something similar for entity integrations: `capability_attributes`. Although that data is actually stored (and restored) in the state machine. 

Can you explain use cases ? From your proposal it sounds like you want to use this in templates?

**Reply by raman325** [reply]
So two use cases:
1. Templates that would allow the user to perform actions on a device using services that require some of this information (e.g. `zwave_js.set_value`)
2. Enhancements to the UI to expose this data to the user which would aid in troubleshooting issues

**Comment by bdraco** [comment]
We have a lot of static data in the state machine that fills up the database, has to be sent over the websocket, serialized, and generally slows everything down just a bit.

Maybe a way to mark a specific attribute as frozen when the entity gets created and store it outside the state machine?

It would also be nice for giant effect_list for lights supported_features, and hvac_modes as storing these in the database probably has little value

**Reply by balloob** [reply]
We have `capability_attributes`. For light it [already contains effect_list](https://github.com/home-assistant/core/blob/dev/homeassistant%2Fcomponents%2Flight%2F__init__.py#L824). However currently capability attributes are stored in the entity registry but also written to the state machine (for backwards compatibility). They were introduced so we could restore entities at startup with these attributes so Google/Homekit/etc sync wouldn't be interrupted.

I wonder if we can build on top of this and:
 - Remove capability attributes from the state machine
 - Allow integrations to define static info that we store in entity registry.

**Reply by raman325** [reply]
I don't totally follow the backwards compatibility piece, but I could explore this option. I think `capability_attributes` is too narrowly scoped of a name to cover all of the attributes I mentioned above, but they may be a good starting place if nothing else.

Re: restoration, I think what I am hearing is that we need to look at allowing the `RestoreEntity` to "restore" data from the entity registry instead of just from the state machine and combine the two? Then we can get rid of the data from the state machine since it's available in the registry

**Reply by raman325** [reply]
Also I really like the idea of marking attributes as frozen. It doesn't solve the extra device attributes piece, but it makes the implementation for an integration owner extremely simple and the complexity can be abstracted away in the core
